File: binary_search/search_2d_matrix/main.cpp

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        for (int r = 0; r < matrix.size(); r++) {
            cout << matrix[r][0] << " " << matrix[r][matrix[r].size() - 1] << endl;
            if (matrix[r][0] == target || matrix[r][matrix[r].size() - 1] == target) {
                return true;
            }

            if (matrix[r][0] < target && matrix[r][matrix[r].size() - 1] > target) {
                int l = 0;
                int h = matrix[r].size() - 1;
                int mid = 0;

                while (l <= h) {
                    mid = l + ((h - l) / 2);

                    if (matrix[r][mid] == target) {
                        return true;
                    }

                    if (matrix[r][mid] > target) {
                        h = mid - 1;
                    } else{
                        l = mid + 1;
                    }
                }
            }
        }

        return false;
    }
};

File: binary_search/search_2d_matrix/main.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        for row in matrix:
            if row[0] == target or row[len(row) - 1] == target:
                return True 

            if row[0] < target < row[len(row) - 1]:
                l = 0
                h = len(row) - 1

                while l <= h:
                    m = l + (h - l) // 2

                    if row[m] == target:
                        return True
                    elif row[m] < target:
                        l = m + 1
                    else:
                        h = m - 1

                return False


        return False
